tool/tsh/common/scan.go

```diff
@@ -0,0 +1,357 @@
+		secretsservice.ClientConfig{
+			ProxyServer: cf.Proxy,
+			Insecure:    cf.InsecureSkipVerify,
+			Log:         slog.Default(),
+		})
+
+	deviceCred, err := native.GetDeviceCredential()
+	if err != nil {
+		return trace.Wrap(err, "device not enrolled")
+	}
+
+	fmt.Fprintf(os.Stdout, "Device trust credentials found.\n")
+	fmt.Fprintf(os.Stdout, "Scanning %s.\n", strings.Join(c.dirs, ", "))
+
+	keys := c.scan(ctx, deviceCred.Id)
+	c.print(os.Stdout, keys)
+
+	fmt.Fprintf(os.Stdout, "Creating Teleport client.\n")
+
+	stream, err := client.ReportSecrets(ctx)
+	if err != nil {
+		// FIXME: handle unimplemented
+		return trace.Wrap(err)
+	}
+
+	cremonyProccess, err := assert.NewCeremony()
+	if err != nil {
+		return trace.Wrap(err, "failed to create ceremony")
+	}
+
+	// Run the device authentication ceremony.
+	// If successful, the device will be authenticated and the device can report its secrets.
+	err = cremonyProccess.Run(
+		ctx,
+		reportToAssertstreamAdapter{stream},
+	)
+	if err != nil {
+		return trace.Wrap(err, "failed to run device authentication ceremony")
+	}
+
+	keysSlice := make([]*accessgraphsecretsv1pb.PrivateKey, 0, len(keys))
+	for _, key := range keys {
+		keysSlice = append(keysSlice, key)
+	}
+
+	if err := stream.Send(&accessgraphsecretsv1pb.ReportSecretsRequest{
+		Payload: &accessgraphsecretsv1pb.ReportSecretsRequest_PrivateKeys{
+			PrivateKeys: &accessgraphsecretsv1pb.ReportPrivateKeys{
+				Keys: keysSlice,
+			},
+		},
+	}); err != nil {
+		return trace.Wrap(err, "failed to send private keys")
+	}
+
+	if err := stream.CloseSend(); err != nil {
+		return trace.Wrap(err, "failed to close send")
+	}
+
+	if _, err := stream.Recv(); err != nil && !errors.Is(err, io.EOF) {
+		return trace.Wrap(err, "error closing the stream")
+	}
+
+	fmt.Fprintf(os.Stdout, "Reported %d SSH fingerprints to Teleport.\n", len(keys))
+
+	return nil
+}
+
+func (c *scanKeysCommand) scan(ctx context.Context, deviceID string) map[string]*accessgraphsecretsv1pb.PrivateKey {
+	allKeys := make(map[string]*accessgraphsecretsv1pb.PrivateKey)
+	for _, dir := range c.dirs {
+		keys := walkAndCheckFiles(ctx, dir, deviceID)
+		maps.Copy(allKeys, keys)
+	}
+
+	return allKeys
+}
+
+func (c *scanKeysCommand) print(out io.Writer, keys map[string]*accessgraphsecretsv1pb.PrivateKey) {
+	if len(keys) == 0 {
+		fmt.Fprintf(out, "No SSH private keys found.\n")
+		return
+	}
+	publicKeyMode := func(mode accessgraphsecretsv1pb.PublicKeyMode) string {
+		switch mode {
+		case accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_PUB_FILE:
+			return "used public key file"
+		case accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_PROTECTED:
+			return "protected private key"
+		case accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_DERIVED:
+			return "derived from private key"
+		default:
+			return "unknown"
+		}
+
+	}
+	fmt.Fprintf(out, "%d SSH private keys found.\n", len(keys))
+	for path, key := range keys {
+		fmt.Fprintf(out, "- SHA256: %q (%s) at %s\n",
```

Comment:
Should we make the message clearer by saying "SHA256 fingerprint" instead of just "SHA256"?